### PR TITLE
rmm: Apply clippy lints

### PR DESCRIPTION
**Pull request by bitboom**
This PR applies clippy lints to rmm related code and CI.

## How to run
You may just run the script.
```sh
./scripts/clippy.sh
```

or execute the command at root.

```sh
cargo clippy --lib -p islet_rmm -- \
	-A clippy::comparison_chain \
	-A clippy::crate_in_macro_def \
	-A clippy::empty_loop \
	-A clippy::explicit_auto_deref \
	-A clippy::from_over_into \
	-A clippy::len_without_is_empty \
	-A clippy::manual_range_contains \
	-A clippy::match_like_matches_macro \
	-A clippy::missing_safety_doc \
	-A clippy::new_without_default \
	-A clippy::type_complexity
```

### CODE REVIEW

File: rmm/src/rsi/mod.rs [hunk 1]
```diff
@@ -358,8 +358,5 @@ pub fn set_event_handler(rsi: &mut RsiHandle) {
 }
 
 fn is_ripas_valid(ripas: u8) -> bool {
-    match ripas as u64 {
-        invalid_ripas::EMPTY | invalid_ripas::RAM => true,
-        _ => false,
-    }
+    matches!(ripas as u64, invalid_ripas::EMPTY | invalid_ripas::RAM)
```
Comment (zpzigi754) [inline rmm/src/rsi/mod.rs:361]: I think that the original one is more readable.

File: rmm/src/rmi/rec/vtcr.rs [hunk 2]
```diff
@@ -13,7 +13,7 @@ pub fn prepare_vtcr(rd: &Rd) -> Result<u64, Error> {
     let ipa_bits = rd.ipa_bits();
 
     // sl0 consists of 2 bits (2^2 == 4)
-    if !(s2_starting_level >= 0 && s2_starting_level <= 3) {
+    if !(0..=3).contains(&s2_starting_level) {
```
Comment (zpzigi754) [inline rmm/src/rmi/rec/vtcr.rs:16]: I think that using the original one is more readable.

File: rmm/src/rmi/rec/mod.rs [hunk 3]
```diff
@@ -133,10 +133,7 @@ impl Rec<'_> {
 
     fn set_runnable(&mut self, flags: u64) {
         const RUNNABLE_OFFSET: u64 = 1;
-        self.runnable = match flags & RUNNABLE_OFFSET {
-            0 => false,
-            _ => true,
-        }
+        self.runnable = !matches!(flags & RUNNABLE_OFFSET, 0);
```
Comment (zpzigi754) [inline rmm/src/rmi/rec/mod.rs:136]: I prefer not using `match!` macro for readability.

File: rmm/src/granule/mod.rs [hunk 4]
```diff
@@ -97,12 +97,12 @@ impl GranuleState {
...
     ($addr:expr) => {{
         {
-            use crate::granule::translation::{GranuleSize, GRANULE_STATUS_TABLE};
-            use crate::granule::validate_addr;
             use vmsa::address::PhysAddr;
             use vmsa::error::Error as MmError;
             use vmsa::page::Page;
             use vmsa::page_table::{self, PageTableMethods};
+            use $crate::granule::translation::{GranuleSize, GRANULE_STATUS_TABLE};
+            use $crate::granule::validate_addr;
```
Comment (zpzigi754) [inline rmm/src/granule/mod.rs:105]: I think that not using `$` looks better for import.

File: lib/vmsa/src/page_table.rs [hunk 5]
```diff
@@ -350,43 +350,44 @@ where
...
+        match L::THIS_LEVEL.cmp(&S::MAP_TABLE_LEVEL) {
+            Ordering::Less => {
+                self.entries[index].set_with_page_table_flags_via_alloc(index, || {
+                    let subtable = unsafe {
+                        Self::alloc_zeroed(
+                            Layout::from_size_align(
+                                L::NextLevel::TABLE_SIZE,
+                                L::NextLevel::TABLE_ALIGN,
+                            )
+                            .unwrap(),
```
Comment (zpzigi754) [inline lib/vmsa/src/page_table.rs:362]: I think that the original one looks better in terms of readability.

**Review by zpzigi754 — COMMENTED** [review]
Good to use clippy!

File: rmm/src/rtt.rs [hunk 6]
```diff
@@ -420,7 +419,7 @@ pub fn data_destroy(id: usize, ipa: usize) -> Result<usize, Error> {
         .ok_or(Error::RmiErrorRtt(0))?
         .into(); //XXX: check this again
 
-    let mut flags = 0 as u64;
+    let mut flags = 0_u64;
```
[the remark this answers is not in the document]
Comment (bitboom) [reply rmm/src/rtt.rs]: In constants case, I agree with you. But `clippy::unnecessary_cast` rule includes unnecessary_cast about the variable case like `ec => Syndrome::Other(ec as u32)`.
So, in constants case, i fixed `0_u64 to 0`

Do you mind keeping this rule?

File: rmm/src/rsi/mod.rs (hunk 1 above)
Comment (bitboom) [reply rmm/src/rsi/mod.rs]: I restored all match exprs :)

File: rmm/src/granule/mod.rs (hunk 4 above)
Comment (bitboom) [reply rmm/src/granule/mod.rs]: I fixed them as you suggested.

File: rmm/src/rmi/rec/vtcr.rs (hunk 2 above)
Comment (bitboom) [reply rmm/src/rmi/rec/vtcr.rs]: I fixed it and add the flag `-A clippy::manual_range_contains`

File: lib/vmsa/src/page_table.rs (hunk 5 above)
Comment (bitboom) [reply lib/vmsa/src/page_table.rs]: I fixed it and `-A clippy::comparison_chain` is added.

File: rmm/src/rsi/mod.rs [hunk 7]
```diff
@@ -42,7 +42,7 @@ pub const ERROR_INPUT: usize = 1;
 pub const ERROR_STATE: usize = 2;
 pub const INCOMPLETE: usize = 3;
 
-pub const VERSION: usize = (1 << 16) | 0;
+pub const VERSION: usize = 1 << 16;
```
Comment (bokdeuk-jeong) [inline rmm/src/rsi/mod.rs:45]: pub const VERSION: usize = (1 << 16) | 0;
This was intended to express the major and minor number.

pub const VERSION: usize = 1 << 16;
Simplifying like this, we are loosing its semantics.  

How about defining const values, VERSION_MAJOR and VERSION_MINOR, for the major and minor as the linux does.
FYI: 
./include/asm/rsi_smc.h
#define RSI_ABI_VERSION_MAJOR           1
:#define RSI_ABI_VERSION_MINOR           0
#define RSI_ABI_VERSION                 ((RSI_ABI_VERSION_MAJOR << 16) |  RSI_ABI_VERSION_MINOR)

File: rmm/src/realm/timer.rs [hunk 8]
```diff
@@ -32,13 +32,13 @@ pub fn restore_state(vcpu: &VCPU<Context>) {
...
-    *&mut timer.cntp_cval_el0 = unsafe { CNTP_CVAL_EL0.get() };
-    *&mut timer.cntp_ctl_el0 = unsafe { CNTP_CTL_EL0.get() };
-    *&mut timer.cnthctl_el2 = unsafe { S3_4_C14_C1_0.get() };
+    timer.cntvoff_el2 = unsafe { CNTVOFF_EL2.get() };
+    timer.cntv_cval_el0 = unsafe { CNTV_CVAL_EL0.get() };
+    timer.cntv_ctl_el0 = unsafe { CNTV_CTL_EL0.get() };
+    timer.cntpoff_el2 = unsafe { S3_4_C14_C0_6.get() }; // CNTPOFF_EL2
+    timer.cntp_cval_el0 = unsafe { CNTP_CVAL_EL0.get() };
+    timer.cntp_ctl_el0 = unsafe { CNTP_CTL_EL0.get() };
+    timer.cnthctl_el2 = unsafe { S3_4_C14_C1_0.get() };
```
Comment (nook1208) [inline rmm/src/realm/timer.rs:41]: It looks more clear !

**Review by nook1208 — APPROVED** [review]
I think clippy is a good tool to clean rust code

File: rmm/src/rsi/mod.rs (hunk 7 above)
Comment (bitboom) [reply rmm/src/rsi/mod.rs]: Thanks for the explanation!

I reflected it like as you suggested.

```rust
const ABI_VERSION_MAJOR: usize = 1;
const ABI_VERSION_MINOR: usize = 0;
pub const VERSION: usize = (ABI_VERSION_MAJOR << 16) | ABI_VERSION_MINOR;
```

File: rmm/src/rtt.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (zpzigi754) [reply rmm/src/rtt.rs]: I see. It seems that there are several cases where using `clippy::unnecessary_cast` is beneficial. Please feel free to keep it.

**Review by bokdeuk-jeong — APPROVED** [review]

**Review by zpzigi754 — APPROVED** [review]

**Comment by hihi-wang** [comment]
Nice to add clippy lints!

**Pull request merged**